### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] Error: Non-null Assertion Failed: Type parameter should be present in scope's variables.

### ISSUE THREAD

**Issue by jleider** [typescript-eslint/typescript-eslint#10012]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://codesandbox.io/p/sandbox/no-unnecessary-type-parameters-bug-6x7w42

### Repro Code

```TypeScript
import type {
  Filterable,
  Filterable1,
  Filterable2,
  Filterable3,
} from "fp-ts/lib/Filterable";
import { constVoid, flow, pipe } from "fp-ts/lib/function";
import type {
  HKT,
  Kind,
  Kind2,
  Kind3,
  URIS,
  URIS2,
  URIS3,
} from "fp-ts/lib/HKT";

type QElement = HTMLElement | SVGElement;

class Q<E extends QElement = QElement> {
  matches(selector: string): boolean {
    return true;
  }

  // This is the line where the linter rule crashes
  static not<F extends URIS3>(F: Filterable3<F>): (selector: string) => <U, L, E_ extends QElement>(elements: Kind3<F, U, L, Q<E_>>) => Kind3<F, U, L, Q<E_>>;
  static not<F extends URIS2>(F: Filterable2<F>): (selector: string) => <L, E_ extends QElement>(elements: Kind2<F, L, E_>) => Kind2<F, L, Q<E_>>;
  static not<F extends URIS>(F: Filterable1<F>): (selector: string) => <E_ extends QElement>(elements: Kind<F, Q<E_>>) => Kind<F, Q<E_>>;
  static not<F_>(F: Filterable<F_>): (selector: string) => <E_ extends QElement>(elements: HKT<F_, Q<E_>>) => HKT<F_, Q<E_>> {
    return (selector: string) => <E_ extends QElement>(elements: HKT<F_, Q<E_>>) => F.filter(elements, (_) => _.matches(selector));
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "baseUrl": ".",
    "downlevelIteration": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "forceConsistentCasingInFileNames": true,
    "incremental": true,
    "jsx": "preserve",
    "lib": [
      "ESNext",
      "dom",
      "dom.iterable",
    ],
    "module": "ESNext",
    "moduleResolution": "node",
    "noEmit": true,
    "noEmitOnError": true,
    "noErrorTruncation": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noPropertyAccessFromIndexSignature": true,
    "noUncheckedIndexedAccess": true,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "preserveConstEnums": true,
    "pretty": true,
    "resolveJsonModule": true,
    "rootDir": ".",
    "sourceMap": true,
    "strict": true,
    "target": "esnext",
    "types": []
  },
  "include": [
    "index.ts"
  ]
}
```

### Expected Result

I would expect the linter to not crash.

### Actual Result

Oops! Something went wrong! :(

ESLint: 8.57.1

Error: Non-null Assertion Failed: Type parameter should be present in scope's variables.
Occurred while linting /Volumes/Sources/web/assets/scripts/dom/q.ts:407
Rule: "@typescript-eslint/no-unnecessary-type-parameters"
    at nullThrows (/Volumes/Sources/node_modules/@typescript-eslint/eslint-plugin/node_modules/@typescript-eslint/utils/dist/eslint-utils/nullThrows.js:19:15)
    at checkNode (/Volumes/Sources/node_modules/@typescript-eslint/eslint-plugin/dist/rules/no-unnecessary-type-parameters.js:55:71)
    at ArrowFunctionExpression[typeParameters], FunctionDeclaration[typeParameters], FunctionExpression[typeParameters], TSCallSignatureDeclaration[typeParameters], TSConstructorType[typeParameters], TSDeclareFunction[typeParameters], TSEmptyBodyFunctionExpression[typeParameters], TSFunctionType[typeParameters], TSMethodSignature[typeParameters] (/Volumes/Sources/node_modules/@typescript-eslint/eslint-plugin/dist/rules/no-unnecessary-type-parameters.js:97:17)
    at ruleErrorHandler (/Volumes/Sources/node_modules/eslint/lib/linter/linter.js:1076:28)
    at /Volumes/Sources/node_modules/eslint/lib/linter/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (/Volumes/Sources/node_modules/eslint/lib/linter/safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (/Volumes/Sources/node_modules/eslint/lib/linter/node-event-generator.js:297:26)
    at NodeEventGenerator.applySelectors (/Volumes/Sources/node_modules/eslint/lib/linter/node-event-generator.js:326:22)
    at NodeEventGenerator.enterNode (/Volumes/Sources/node_modules/eslint/lib/linter/node-event-generator.js:340:14)

### Additional Info

The eslint sandbox doesn't seem to support importing packages and this failure depends on a 3rd party library's (fp-ts) types. Please let me know if importing modules is supported.

I also tried to set it up with codesandbox.io but I don't think they support running eslint as far as I can tell so I'm not sure if this will be reproduceable outside of our codebase. 

Hopefully that code snippet will be enough to pinpoint the issue.

**Comment by kirkwaiblinger** [comment]
Hi @jleider, thanks for reporting this! 
As you've noted, this is a complex case that depends on third party libraries. It would help us address this faster if you could provide a clonable minimal reproduction repro, like in our ["complex bug" template](https://github.com/typescript-eslint/typescript-eslint/issues/new?assignees=&labels=bug%2Ctriage&projects=&template=08-bug-report-complex.yaml&title=Bug%3A+%3Cshort+description+of+the+issue%3E). Thanks!

**Comment by kirkwaiblinger** [comment]
fwiw - i tried downloading the codesandbox example and i wasn't seeing the described issue. 

I'm going to close for now on the speculation that there's just a dependency version issue in the repo in which you have observed this error. But we can definitely reopen if you're able to come up with a repro! 🙂
